torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -234,8 +236,64 @@ inline void errorIfCapturingNonCapturableNCCL() {
   }
 }
 
+void markRankNotStuck(char *rankStatusHelper, int rank) {
+  if (!rankStatusHelper) {
+    LOG(ERROR) << "rankStatusHelper is NULL";
+    return;
+  }
+  rankStatusHelper[rank] = 0x00;
+}
+
+void markRankStuck(char *rankStatusHelper, int rank) {
+  if (!rankStatusHelper) {
+    LOG(ERROR) << "rankStatusHelper is NULL";
+    return;
+  }
+  rankStatusHelper[rank] = 0x01;
+}
+
+void markRankTimedOut(char *rankStatusHelper, int rank) {
+  if (!rankStatusHelper) {
+    LOG(ERROR) << "rankStatusHelper is NULL";
+    return;
+  }
+  rankStatusHelper[rank] = 0x02;
+}
+
 } // namespace
 
+RankStatusHelper::RankStatusHelper(size_t worldSize): worldSize_(worldSize) {
+  // assume file exists
+  int fd;
+  if ((fd = open("rank_statuses", O_RDWR, 0)) == -1) {
+    LOG(ERROR) << "unable to open rank_statuses";
+  }
+  handle_ = static_cast<char*>(mmap(
```

Comment:
nit: uint8_t*?

Reply:
Is this guaranteed to initialize the entire file with 0x00? If not should we explicitly fill the buffer with 0x00?

Reply:
I initialize the file with `0xFF` in `_create_rank_statuses_file` in `spawn.py`